[JIT] Add determinism checks for ops in SchemaInfo subclass

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81352
* #81007
* #81002
* __->__ #81000
* #80984
* #80972
* #80918

- Added is_non_deterministic which returns whether a given op is non-deterministic. Currently this is implemented with a hard-coded list of non-deterministic functions copied from ir.cpp in AliasDB, but this will eventually be implemented by returning with a given schema has the non_deterministic tag. 
- Tested is_non_deterministic method with a deterministic op and a non deterministic op in test_schema_info.cpp

**Note that the case for op "aten::dropout(Tensor input, float p, bool train) -> Tensor" which is deterministic whenever "train=false" is not accounted for in this pr and will be fixed in a later pr. Currently "aten::dropout(Tensor input, float p, bool train) -> Tensor" is always considered nondeterministic.**

Commit:
Add determinism checks for ops in SchemaInfo subclass

[ghstack-poisoned]

Commit:
Update on "[JIT] Add determinism checks for ops in SchemaInfo subclass"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add determinism checks for ops in SchemaInfo subclass"




[ghstack-poisoned]

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -59,6 +59,39 @@ bool SchemaInfo::is_mutable(c10::string_view name) {
   return is_mutable(*index);
 }
 
+bool SchemaInfo::is_non_deterministic() const {
```
add a comment that this came from JIT (note: it probably makes sense to put the comment in the cpp file cause that's where the implementation is: i.e., I wouldn't wonder where this came from just by looking at the function signature in the .h file, but in the .cpp file I might see this list of ops and wonder where it came from)

davidberard98 replied:
nit: i'd probably name it `is_nondeterministic` for consistency, unless there's a reason to prefer `non_deterministic`

davidberard98 reviewed (commented):
nice, looks good! just a few minor comments

Commit:
Update on "[JIT] Add determinism checks for ops in SchemaInfo subclass"


- Added is_non_deterministic which returns whether a given op is non-deterministic. Currently this is implemented with a hard-coded list of non-deterministic functions copied from ir.cpp in AliasDB, but this will eventually be implemented by returning with a given schema has the non_deterministic tag. 
- Tested is_non_deterministic method with a deterministic op and a non deterministic op in test_schema_info.cpp

**Note that the case for op "aten::dropout(Tensor input, float p, bool train) -> Tensor" which is deterministic whenever "train=false" is not accounted for in this pr and will be fixed in a later pr. Currently "aten::dropout(Tensor input, float p, bool train) -> Tensor" is always considered nondeterministic.**

[ghstack-poisoned]

Commit:
Update on "[JIT] Add determinism checks for ops in SchemaInfo subclass"


- Added is_non_deterministic which returns whether a given op is non-deterministic. Currently this is implemented with a hard-coded list of non-deterministic functions copied from ir.cpp in AliasDB, but this will eventually be implemented by returning with a given schema has the non_deterministic tag. 
- Tested is_non_deterministic method with a deterministic op and a non deterministic op in test_schema_info.cpp

**Note that the case for op "aten::dropout(Tensor input, float p, bool train) -> Tensor" which is deterministic whenever "train=false" is not accounted for in this pr and will be fixed in a later pr. Currently "aten::dropout(Tensor input, float p, bool train) -> Tensor" is always considered nondeterministic.**

[ghstack-poisoned]

davidberard98 reviewed (approved):
looks good!!

goldenxuett commented:
@pytorchbot merge
